packages/lodestar/src/sync/regular/oneRangeAhead/fetcher.ts

```diff
@@ -0,0 +1,106 @@
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {SignedBeaconBlock, Slot} from "@chainsafe/lodestar-types";
+import {ErrorAborted, ILogger, sleep} from "@chainsafe/lodestar-utils";
+import deepmerge from "deepmerge";
+import PeerId from "peer-id";
+import {IRegularSyncModules} from "..";
+import {defaultOptions, IRegularSyncOptions} from "../options";
+import {IBeaconChain} from "../../../chain";
+import {INetwork} from "../../../network";
+import {getBlockRange} from "../../utils/blocks";
+import {ISlotRange, ISyncCheckpoint} from "../../interface";
+import {ZERO_HASH} from "../../../constants";
+import {IBlockRangeFetcher} from "./interface";
+
+export class BlockRangeFetcher implements IBlockRangeFetcher {
+  protected readonly config: IBeaconConfig;
+  private readonly network: INetwork;
+  private readonly chain: IBeaconChain;
+  private readonly logger: ILogger;
+  private readonly opts: IRegularSyncOptions;
+  // for control range across next() calls
+  private lastFetchCheckpoint: ISyncCheckpoint;
+  // for each next() call
+  private rangeStart: Slot = 0;
+  private rangeEnd: Slot = 0;
+  private getPeers: () => Promise<PeerId[]>;
+
+  constructor(options: Partial<IRegularSyncOptions>, modules: IRegularSyncModules, getPeers: () => Promise<PeerId[]>) {
+    this.config = modules.config;
+    this.network = modules.network;
+    this.chain = modules.chain;
+    this.logger = modules.logger;
+    this.opts = deepmerge(defaultOptions, options);
+    this.getPeers = getPeers;
+    this.lastFetchCheckpoint = {blockRoot: ZERO_HASH, slot: 0};
+  }
+
+  public setLastProcessedBlock(lastProcessedBlock: ISyncCheckpoint): void {
+    this.lastFetchCheckpoint = lastProcessedBlock;
+  }
+
+  /**
+   * Get next block range.
+   */
+  public async next(): Promise<SignedBeaconBlock[]> {
+    this.updateNextRange();
+    let result: SignedBeaconBlock[] | null = null;
+    while (!result || !result!.length) {
+      let slotRange: ISlotRange | null = null;
+      try {
+        const peers = await this.getPeers();
+        if (result && !result.length) await this.handleEmptyRange(peers);
+        slotRange = {start: this.rangeStart, end: this.rangeEnd};
+        result = await getBlockRange(this.logger, this.network.reqResp, peers, slotRange);
+      } catch (e) {
+        this.logger.debug("Regular Sync: Failed to get block range ", {...(slotRange ?? {}), error: e.message});
```

Comment:
I think Javascript allows spreading nullish values
```
{...null, ...undefined}
```
Also, give the entire error to the logger, it will handle the formating